### ISSUE TITLE
Threads: Exit Beta

### ISSUE THREAD

**Issue by daniellekirkwood** [vector-im/element-meta#187]
There are items that we did not deliver as a part of the MVP for Threads that need to be delivered as soon after the MVP as possible.

[Project board](https://github.com/orgs/vector-im/projects/17/views/49)

## Proof of concepts

- \[x\] https://github.com/vector-im/element-web/issues/22980
- \[x\] https://github.com/vector-im/element-web/issues/22981

## Matrix Spec Changes

- \[x\] https://github.com/matrix-org/matrix-spec-proposals/pull/3771
~~https://github.com/matrix-org/matrix-spec-proposals/pull/3772~~ This has been removed from Exit Beta as per [Notification requirements](https://docs.google.com/document/d/1DYyDsf24OlrME3h-Hg-H2GvCrKkqZJrQwIbNWjqxB2A/edit?usp=sharing)
- \[x\] https://github.com/matrix-org/matrix-spec-proposals/pull/3773
- \[x\] https://github.com/matrix-org/matrix-spec-proposals/pull/3856

## User Improvements

The first beta launch of Threads was successful but missed the mark in some areas. Before removing the Beta label we should resolve the following user issues:

- **[Threads: Notifications](https://docs.google.com/document/d/1DYyDsf24OlrME3h-Hg-H2GvCrKkqZJrQwIbNWjqxB2A/edit?usp=sharing)**
	- Notifications should be predictable. They should be reliable cross-platform and enable users to understand which spaces, rooms, and threads contain messages they have not yet read.
- **Thread list sorting and filtering**
	- The Thread List Panel inside a room should be ordered by "most recently responded to" so that active threads are nearer the top \(even if the root message is further back in time\).
	- The Thread List Panel should also be filterable by Threads that the user has Participated in, so that they can see the threads they either started or responded to.
- **Scrolling and loading**
	- When a room contains a lot of Threaded messages, the timeline struggles to scroll back smoothly, even when the Thread List Panel is closed. The Thread List Panel also struggles to load in a timely manner and we should be aiming to improve the performance of our product.
- **Fallback for incomplete Threads**
	- There may be several reasons that a Thread in the Thread Timeline is incomplete: the user doesn't have permission to the history, or the room was upgraded, etc. We have designs for these use cases and we should implement them where possible.

## Functional Improvements

- \[x\] https://github.com/vector-im/element-meta/issues/196
- \[x\] https://github.com/vector-im/element-meta/issues/195
- \[x\] https://github.com/vector-im/element-meta/issues/235

Stretch Goal:
- \[ \] https://github.com/vector-im/element-meta/issues/240

## Design Improvements

- \[x\] https://github.com/vector-im/element-web/issues/21502 \(Need to make a decision regarding BACAT scrolling\)
- \[ \] https://github.com/vector-im/element-android/issues/5593
- \[x\] https://github.com/vector-im/element-ios/issues/5878
- \[ \] https://github.com/vector-im/element-meta/issues/121
- \[ \] https://github.com/vector-im/element-ios/issues/5396
- \[ \] https://github.com/vector-im/element-meta/issues/1925
- \[x\] https://github.com/vector-im/element-web/issues/21759
- \[x\] https://github.com/vector-im/element-web/issues/21335
- \[x\] https://github.com/vector-im/element-web/issues/20908
- \[x\] https://github.com/vector-im/element-android/issues/5151 Implemented by Aris but needs some improvements.
- \[x\] https://github.com/vector-im/element-ios/issues/5502

## Major Bugs

As part of this launch we are also committing to resolving all P1 issues

- Web: https://github.com/vector-im/element-web/issues?q=is%3Aissue+is%3Aopen+label%3Aa-threads+label%3As-major
- iOS: https://github.com/vector-im/element-ios/issues?q=is%3Aissue+is%3Aopen+label%3Aa-threads+label%3As-major
- Android: https://github.com/vector-im/element-android/issues?q=is%3Aissue+is%3Aopen+label%3Aa-threads+label%3As-major

Critical bugs:
- Web: https://github.com/vector-im/element-web/issues?q=is%3Aissue+is%3Aopen+label%3Aa-threads+label%3As-critical
- iOS: https://github.com/vector-im/element-ios/issues?q=is%3Aissue+is%3Aopen+label%3Aa-threads+label%3As-critical
- Android: https://github.com/vector-im/element-android/issues?q=is%3Aissue+is%3Aopen+label%3Aa-threads+label%3As-critical

## Release

- \[x\] https://github.com/vector-im/element-meta/issues/566
- \[x\] https://github.com/vector-im/element-ios/issues/5799
- \[x\] https://github.com/vector-im/element-android/issues/5503
- \[x\] https://github.com/vector-im/element-web/issues/21910

**Comment by janogarcia** [comment]
## 🐛 Bugs

### Design implementation review

- **Web**: Incomplete tasks. <br> https://github.com/vector-im/element-web/issues/21502 <br> Exceptions: "Align thread list to the top" and "Align timeline to the top of the timeline container." which are blocked by BACAT (Bottom-Aligned, Clipped-At-Top) scrolling. More on that below, in the "✋ Blocked" section.
- **Android**: Uncompleted tasks + New tasks added for  1.4.10 RC <br> https://github.com/vector-im/element-android/issues/5593
- **iOS**: Uncompleted tasks + New tasks added for 1.8.12 RC <br> https://github.com/vector-im/element-ios/issues/5878

### Thread notifications

Please refer to relevant thread:

🔔 Notifications: MVP status and post-MVP plans.
https://matrix.to/#/!YFprWIqHBfwwwCBYOe:matrix.org/$WUveEsZH-KliepTx6iyWnLBV9vDiDVGajOkr7pMjJ3o

From all the stuff marked as "🔮 Post MVP" I'd only recommend implementing the following ones for the "Exit Beta" milestone:

- **Thread list**: Sync the read status across clients.
- **Room header**: Display unread count in the room header badge.
- **Thread timeline**: Only mark thread as read if the user actually read it (went past the unread marker down to the latest message, rather than marking it read as soon as the user opens the thread).

Everything else marked as "🔮 Post MVP" but excluded for the "Exit Beta" milestone can be reconsidered at a later stage. For example, potential Post v1 enhancements:

- **Room list: Thread notifications in room list**. Help users keep track of thread updates across rooms and navigating them, instead of having to peek into every room and look for thread updates. <br> → [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=3586%3A328211)
- **Room timeline: Thread events**. Help surfacing relevant thread activity that otherwise would go unnoticed (e.g., a user starts a thread on an old message). <br> → [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=1246%3A231502)

### Thread notifications: Thread replies to root messages going unnoticed

Please refer to relevant thread:

🎛 On thread filtering params
https://matrix.to/#/!YFprWIqHBfwwwCBYOe:matrix.org/$owxab8NKxIqw2sMoypkxnGiqDUpc282o8VYAQwzrFUE

Current implementation prevents you from being notified when someone starts a thread on one of your messages without mentioning or quoting you.

“My threads” should include threads where I’m the author of the root message even if I didn’t reply to it or no one mentioned me directly.

Failing to implement it means users will be missing replies to their messages.

- User A post a message in the main timeline.
- User B starts a thread for User A message, without directly mentioning him (e.g., User B message is `That looks promising. Any chance I can test the service before I commit to your contract?`)
- User A never gets a notification for User B request.
- Missed business opportunity.

### Rendering of content types and formatting styles in root messages and thread summaries

I've recreated all the test cases, run them again and updated the status for all platforms (Apr.

Related issues:

- https://github.com/vector-im/element-meta/issues/121

### Handling incomplete threads

The user can’t view messages in a thread prior to her join or invite date, due to room permissions. Or, the root message is missing due to Federation (delayed or permanently unavailable) or some other reason (buggy or malicious server).

https://github.com/vector-im/element-meta/issues/1925

Design explorations:

- Web: [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=3264%3A320839)
- Android: [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=3267%3A335887)
- iOS: [Figma](https://www.figma.com/file/T309ztx0sNyOOK6NKVLHsK/Threads?node-id=3267%3A338027)

### Handling room upgrades

Or how we could minimize the negative impact of room upgrades on Threads. Room upgrades are rather infrequent, but there will probably be new room versions released in 2022.

Related issues:

- https://github.com/vector-im/element-meta/issues/61#issuecomment-1014629516

## ✨ Improvements

### Less ambiguous rendering of root messages and thread summaries pairs in the main timeline

- **Web**: Implemented by Robin but needs some improvements. <br> https://github.com/vector-im/element-web/issues/20908 
- **Android**: Implemented by Aris but needs some improvements. <br> https://github.com/vector-im/element-android/issues/5151
- **iOS**: ✋ Potentially blocked by the complexity of introducing changes to the timeline component on iOS. <br>  https://github.com/vector-im/element-ios/issues/5502

### Implement improved spacing for the thread list and timeline

Related issues:

- https://github.com/vector-im/element-web/issues/21759

### Load states for thread panel (thread list and thread view)

Related issues:

- https://github.com/vector-im/element-web/issues/21335

## ✋ Blocked

### Persist scrolling position of thread list

Losing scroll position while navigating threads makes you easily lose context which is quite inconvenient and adds friction for example when catching up with updates/unreads. The user should be able to continue browsing the list at their previous scrolling position.

Currently blocked “until we have a paginated API for this, as the solution will be rather different”.

Related issues:

- https://github.com/vector-im/element-web/issues/21735

### Sort order of thread list on Web

We should be able to order items from most recent activity on top. Currently blocked by BACAT (Bottom-Aligned, Clipped-At-Top) infinite scrolling implementation.

- https://github.com/vector-im/element-web/issues/21502#issuecomment-1080708327
- [matrix-react-sdk/docs/scrolling.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/docs/scrolling.md#bacat-bottom-aligned-clipped-at-top-scrolling)

Related issues:

- https://github.com/vector-im/element-web/issues/19393
- https://github.com/vector-im/element-web/issues/20654#issuecomment-1081638331

## ✅ Already fixed

- **Populating the thread list on tab reload** <br> https://github.com/vector-im/element-web/issues/21737
- **Horizontal alignment for media and reaction containers in threads is off** <br> https://github.com/vector-im/element-web/issues/21780
- **Persist thread open state** <br> https://github.com/vector-im/element-web/issues/21741

**Comment by janogarcia** [comment]
@daniellekirkwood @novocaine I basically reviewed every source of information I had on threads for compiling the list above:

- Open issues across all platforms (GitHub)
- Feature ideas backlog (GitLab, GitHub...)
- Design explorations backlog (Figma)
- Personal notes

I tried to figure out what needs to be accomplished to exit threads from beta while keeping potential future enhancements as a separate concern.

**Comment by janogarcia** [comment]
@novocaine 🐛 We need to log an issue somewhere (probably as a meta issue) to describe the current limitation where a room with many threads fails to load message history in the main timeline when scrolling back. 

Related discussion in Threads Delivery room:
https://matrix.to/#/!YFprWIqHBfwwwCBYOe:matrix.org/$RQxl9muD26bpmpJNfF3LSfgyNvg0oxGXiVc3c3Q8Mg8

**Comment by novocaine** [comment]
> @novocaine 🐛 We need to log an issue somewhere (probably as a meta issue) to describe the current limitation where a room with many threads fails to load message history in the main timeline when scrolling back.
> 
> Related discussion in Threads Delivery room: https://matrix.to/#/!YFprWIqHBfwwwCBYOe:matrix.org/$RQxl9muD26bpmpJNfF3LSfgyNvg0oxGXiVc3c3Q8Mg8

Michael has added this to the scope already https://github.com/vector-im/element-meta/issues/240

**Comment by janogarcia** [comment]
Awesome, thanks!

**Comment by janogarcia** [comment]
@gsouquet @daniellekirkwood Would it be possible to move the following issue from [my list above](https://github.com/vector-im/element-meta/issues/187#issuecomment-1106680814) to the Exit Beta epic?

Losing scroll position on Web when navigating between the thread list and thread detail views:

As noted below, it was blocked on a technical constraint the last time we discussed it. If that's not longer a limitation we should consider including it for the epic, otherwise the experience when navigating back and forth between the Thread List and the Thread view will be subpar. 

Scenario:

1. Imagine that you're in your email client of choice and have scrolled down the list in your Inbox (or Archive);
2. then you click or tap on an email to view it, and;
3. then you navigate back to the Inbox only to see that you are sent back to the initial/top scroll position of your Inbox, instead of going back to the scroll position you were before.

---

### Persist scrolling position of thread list

Losing scroll position while navigating threads makes you easily lose context which is quite inconvenient and adds friction for example when catching up with updates/unreads. The user should be able to continue browsing the list at their previous scrolling position.

Currently blocked “until we have a paginated API for this, as the solution will be rather different”.

Related issues:

- https://github.com/vector-im/element-web/issues/21735

**Comment by daniellekirkwood** [comment]
We want to make sure we do not _exclude_ the possibility of a follow/subscribe feature in the future.

**Comment by daniellekirkwood** [comment]
Today we made the decision to add an icon to a message preview when that message is from a thread. This behaviour of showing message previews in threads has been disjointed on all the platforms - this approach unifies it. 

GH issues are being created at a platform level to reflect it and build the changes :)

**Comment by clokep** [comment]
This is a follow-on to vector-im/element-web#203 FTR.

**Comment by daniellekirkwood** [comment]
This epic will be closed as the majority of work has been completed and the target for the quarter achieved. 

The outstanding items are being tracked in their issues/repo's _or_ have been deprioritised due to resourcing shifts.
